Fixed url to file conversion breaking paths with spaces

dmatej opened the pull request:
This PR fixes part of issues seen when manually testing another branch on Windows using path with spaces.
Fixes also a validation issue when URL with empty path could pass into the classloader.
I have cherrypicked also another commit refactoring test which I needed here to confirm issues we have noticed in sources.

dmatej committed:
Simplified CommonClassLoaderServiceImplTest - using HK2

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed path resolution in CommonClassLoaderServiceImpl

- Original code broke when path contained spaces
- Added validation of the path before it was added anywhere.
  - Note that FileUtils.toFile throws IAE when the path is empty too, however
    this code is now probably unreachable because of the validation.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added conversion method URL -> File

- This conversion is used very often and I got tired of all issues with that.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

pzygielo commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -172,12 +172,7 @@ public String getCommonClassPath() {
-                .map(File::new)
-                .map(File::getAbsolutePath)
-                .collect(Collectors.joining(File.pathSeparator));
+        return urls.map(FileUtils::toFile).map(File::getAbsolutePath).collect(Collectors.joining(File.pathSeparator));
```
I'd keep original line breaks to show that only part of the method changed, not the whole one.

dmatej replied:
It is under 120 characters border, so it is not so hard to read. I would agree otherwise.

OndroMih commented on nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java:
```diff
@@ -1032,6 +1033,37 @@ public static File[] findFilesInDir(File dir, final String regexp) {
+     * @throws IllegalArgumentException if the URL cannot be converted to a File
+     *
+     */
+    public static String toAbsolutePath(URL resourceUrl) throws IllegalArgumentException {
```
This method is not used. And that's correct - I would always prefer to explicitly call `toFile(URL).getPath()` for clarify.
Can we delete this method?

dmatej replied:
It is used in 7.1 branch, this is backport. But ok, I will rebase that later and remove it there too.

OndroMih commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -172,12 +172,7 @@ public String getCommonClassPath() {
-                .map(File::new)
-                .map(File::getAbsolutePath)
-                .collect(Collectors.joining(File.pathSeparator));
+        return urls.map(FileUtils::toFile).map(File::getAbsolutePath).collect(Collectors.joining(File.pathSeparator));
```
Where did `.filter(Predicate.not(String::isBlank))` disappear? Why was it even there? Is it OK to remove it or it's possible that some urls are blank?
Maybe it's worth adding `.filter(File::exists)` ?

OndroMih replied:
Also note that `new File("").getAbsolutePath()` gives the value of the home directory. So I suggest adding also `.filter( f -> !f.getPath().isBlank())`

dmatej replied:
Hmm, I see it is used by the addLibraryCommand, so you are right and that resolves even Piotrek's suggestion, because it will be longer line.  I still have doubts it is possible to get an empty string here, maybe would be useful to cover that by a test.

OndroMih replied:
`addLibraryCommand` verifies that the file exists and only then it adds the URL. So it seems that urls cannot be empty in any case and my comment is void.

dmatej replied:
Funny, I extended tests and found much more old and dangerous bugs ... will be fixed now.

dmatej committed:
Improved CommonClassLoaderServiceImplTest

- Added coverage of paths with spaces
- Added coverage of adding an empty file path

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>
